src/main/java/org/sonatype/nexus/ci/nxrm/v3/MoveComponentsStep.java

```diff
@@ -0,0 +1,138 @@
+import static org.sonatype.nexus.ci.nxrm.Messages.MoveComponents_Validation_DestinationRequired;
+import static org.sonatype.nexus.ci.util.RepositoryManagerClientUtil.nexus3Client;
+
+public class MoveComponentsStep
+    extends Builder
+    implements SimpleBuildStep
+{
+  private final String nexusInstanceId;
+
+  private final String tagName;
+
+  private final String destination;
+
+  @DataBoundConstructor
+  public MoveComponentsStep(final String nexusInstanceId,  final String tagName, final String destination)
+  {
+    this.nexusInstanceId = checkArgument(nexusInstanceId, isNotBlank(nexusInstanceId),
+        Common_Validation_NexusInstanceIDRequired());
+    this.destination = checkArgument(destination, isNotBlank(destination),
+        MoveComponents_Validation_DestinationRequired());
+    this.tagName = checkArgument(tagName, isNotBlank(tagName), Common_Validation_Staging_TagNameRequired());
+  }
+
+  public String getNexusInstanceId() {
+    return nexusInstanceId;
+  }
+
+  public String getTagName() {
+    return tagName;
+  }
+
+  public String getDestination() {
+    return destination;
+  }
+
+  @Override
+  public void perform(@Nonnull final Run run, @Nonnull final FilePath workspace, @Nonnull final Launcher launcher,
+                      @Nonnull final TaskListener listener) throws InterruptedException, IOException
+  {
+    try {
+      RepositoryManagerV3Client client = nexus3Client(nexusInstanceId);
+      List<ComponentInfo> components = client.move(destination, tagName);
+      listener.getLogger().println("Move successful. Destination: '" + destination + "' Components moved:\n" +
+          components.stream()
+          .map(c -> c.getGroup() + ":" + c.getName() + ":" + c.getVersion())
+          .collect(joining("\n")));
+    }
+    catch (RepositoryManagerException e) {
+      listener.getLogger().println("Failing build due to: " + e.getResponseMessage().orElse(e.getMessage()));
+      run.setResult(FAILURE);
+    }
+  }
+
+  @Extension
+  @Symbol("moveComponents")
+  public static final class DescriptorImpl
+      extends BuildStepDescriptor<Builder>
+  {
+    @Override
+    public String getDisplayName() {
+      return MoveComponents_DisplayName();
+    }
+
+    @Override
+    public boolean isApplicable(final Class<? extends AbstractProject> jobType) {
+      return true;
+    }
+
+    public FormValidation doCheckNexusInstanceId(@QueryParameter String value) {
+      return NxrmUtil.doCheckNexusInstanceId(value);
+    }
+
+    public ListBoxModel doFillNexusInstanceIdItems() {
+      return NxrmUtil.doFillNexusInstanceIdItems(NxrmVersion.NEXUS_3);
+    }
+
+    public FormValidation doCheckDestination(@QueryParameter String value) {
+      return NxrmUtil.doCheckNexusRepositoryId(value);
+    }
+
+    public ListBoxModel doFillDestinationItems(@QueryParameter String nexusInstanceId) {
+      return NxrmUtil.doFillNexusRepositoryIdItems(nexusInstanceId);
```

Comment:
So to date this utility was only used for the publisher step, which is restricted (due to NX2) to only maven hosted repos. I don't believe we want that restriction on move, so this will need to add support for retrieving all supported hosted repositories.

Reply:
arg...right right. Correct, I'll look at that.

Reply:
@wrightkris quick question on this...It looks like the `getApplicableRepositories` in Nxrm3Util is restricting to hosted maven repositories. I don't recall, are we only claiming (for right now) upload support for maven in the jenkins plugin? I guess I'm wondering if the `getApplicableRepositories` method should be adjusted to also support move than just maven for uploads (maybe there was a reason that wasn't done?)...and maybe that's what you meant when you said 'add support'.  

Otherwise, I can add an additional method for retrieving all repositories to support the move operation.

Reply:
As for move, I don't think we're restricted other than attempting to move items across formats (which would yield an error), so my thought was to add a new method to get all repos, or modify the existing method to require a format parameter.